mlir/lib/Dialect/Vector/IR/VectorOps.cpp

```diff
@@ -5643,32 +5643,24 @@ OpFoldResult ShapeCastOp::fold(FoldAdaptor adaptor) {
       return bcastOp.getSource();
   }
 
+  // Replace shape_cast(arith.constant) with arith.constant. Currently only
+  // handles splat constants.
+  if (auto constantOp = getSource().getDefiningOp<arith::ConstantOp>()) {
+    if (auto dense = llvm::dyn_cast<SplatElementsAttr>(constantOp.getValue())) {
+      return DenseElementsAttr::get(cast<VectorType>(getType()),
+                                    dense.getSplatValue<Attribute>());
+    }
+  }
+
+  // Replace shape_cast(poison) with poison.
+  if (getSource().getDefiningOp<ub::PoisonOp>()) {
+    return ub::PoisonAttr::get(getContext());
+  }
```

Comment:
You don't need to check getDefiningOp, the adaptor will automatically give you an attribute if it can. Instead, you can do:

if (auto dense = llvm::dyn_cast<SplatElementsAttr>(adaptor.getSource()) {
...
}

Same for ub::PoisonOp, but use ub::PosionAttr

Reply:
That is simpler. Note I needed to `if (auto dense = llvm::dyn_cast_if_present(adaptor.getSource())`  -- the '_if_present' for the case where the shape cast is the source is not already a constant/attribute.